tutorial/lesson_22_jit_performance.cpp

```diff
@@ -0,0 +1,174 @@
+// LD_LIBRARY_PATH=<path/to/libHalide.so> ./lesson_20
+
+// On os x:
+// g++ lesson_22*.cpp -g -I <path/to/Halide.h> -L <path/to/libHalide.so> -lHalide -o lesson_22 -std=c++17
+// DYLD_LIBRARY_PATH=<path/to/libHalide.dylib> ./lesson_22
+
+// If you have the entire Halide source tree, you can also build it by
+// running:
+//    make tutorial_lesson_22_jit_performance
+// in a shell at the top of the halide source tree.
+
+#include "Halide.h"
+#include "halide_benchmark.h"
+#include <stdio.h>
+
+using namespace Halide;
+using namespace Halide::Tools; // for benchmark()
+
+// Lets define a helper function to construct a simple pipeline that we'll use for our performance tests
+Pipeline make_pipeline() {
+
+    // We'll start with a simple transpose operation 
+    Func input("input"), output("output");
+    Var x("x"), y("y");
+
+    // Fill the input with a linear combination of the coordinate values
+    input(x, y) = cast<uint16_t>(x + y);
+    input.compute_root();
+
+    // Transpose the rows and cols 
+    output(x, y) = input(y, x);
+
+    // Schedule it ... there's a number of possibilities here to do an efficient block-wise transpose
+    Var xi("xi"), yi("yi");
+    
+    // Let's focus on 8x8 subtiles, and then vectorize across X, and unroll across Y
+    output.tile(x, y, xi, yi, 8, 8).vectorize(xi).unroll(yi);
+
+    // For more advanced scheduling: 
+    //
+    // We can improve this even more by using the .in() directive (see Tutorial 19), 
+    // which allows us to interpose new Funcs in between input and output.
+    // 
+    // Here we can inject a block_transpose function to allow us to do 8 vectorized loads from the input.
+    // Func block_transpose("block_transpose"), block("block");
+    // block_transpose = input.in(output).compute_at(output, x).vectorize(x).unroll(y);
+    //
+    // And now lets reorder and vectorize in X across the block 
+    // block = block_transpose.in(output).reorder_storage(y, x).compute_at(output, x).vectorize(x).unroll(y);
+
+    return Pipeline(output);
+}
+
+int main(int argc, char **argv) {
+
+    
+    // Lets measure the performance of calling realize on a simple pipeline
+    {
+        Pipeline pipeline = make_pipeline();
+
+        // Create an 1024x1024 output buffer to hold the results 
+        Buffer<uint16_t> result(1024, 1024);
+
+        // Lets realize the pipeline (which implicitly JIT compiles, dynamically links and executes the code)
+        size_t count = 0;
+        double t = benchmark([&]() {
+            pipeline.realize(result);
+            ++count;
+        });
+
+        std::cout << "Execute Pipeline (realize): " << int(count / t) << " times/sec\n";
+    }
+
+    // This time let's explicitly invoke the JIT compile, before we realize and execute the pipeline
+    {
+        Pipeline pipeline = make_pipeline();
+        Buffer<uint16_t> result(1024, 1024);
+
+        // Let's explicitly JIT compile before we realize, which will cache the generated code
```

Comment:
Suggestion: add comments indicating the rough expected improvement on a typical desktop system